DietPi-Software | Add folding@home to software list

MichaIng opened the pull request:
**Status**: WIP
- [x] Software category?
- [x] DietPi-Services handling?
- [x] Allow fully automated install by providing user name, passkey and power choice via dietpi.txt?
- [x] Move data dir to dietpi_userdata?
- [x] Changelog entry

**Reference**: https://github.com/Fourdee/DietPi/issues/1985

**Commit list/description**:
+ DietPi-Software | Add folding@home to software list
  - Preconfigured autostart + DietPi team
  - User name, passkey, power choice need to be entered

MichaIng committed:
v6.13

+ DietPi-Software | Add folding@home to software list

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -1514,6 +1514,24 @@ _EOF_
 	   aSOFTWARE_REQUIRES_JAVA_JRE_JDK[$index_current]=1
 			 aSOFTWARE_ONLINEDOC_URL[$index_current]='f=8&t=5&p=6202#p6202'
 
```
@Fourdee 
Watching alignment of the software arrays, I guess your local code editor handled tabs differently than github (tab = 8 digits space)?
Found my notepad++ with default 2 spaces, if I remember right, so raw downloads from github looked wrong at first.

Shall we align all of them to look pretty on github... by times... ?

Fourdee replied:
@MichaIng 

I use notepad++ with default tab size of 4
![image](https://user-images.githubusercontent.com/7077026/43685002-f37c32a6-98a2-11e8-86a5-ece1666b20e6.png)

![image](https://user-images.githubusercontent.com/7077026/43685010-0ad416da-98a3-11e8-91f4-fa6f90f491b7.png)

> Shall we align all of them to look pretty on github... by times... ?

Yep, we probably should do that (at some point, minor). Apparently GitHub default is 8 for a tab:
https://github.com/lukechilds/github-custom-tab-size#github-custom-tab-size

MichaIng replied:
@Fourdee 
Jep, it depends on the file ending as well, AFAIK. For .txt/.md files I remember a default of `<tab>` being translated into 2 times `<blank>`. But at least all our script files have tab -> tab with 8 blank width.

Ah now I see you format intention, having all `=` aligned.

Fourdee commented on dietpi/dietpi-software:
```diff
@@ -1514,6 +1514,24 @@ _EOF_
+		aSOFTWARE_WHIP_DESC[$index_current]='distributed disease research project'
+		aSOFTWARE_CATEGORY_INDEX[$index_current]=6
+		aSOFTWARE_TYPE[$index_current]=0
+		aSOFTWARE_ONLINEDOC_URL[$index_current]=''
```
Created:
https://dietpi.com/phpbb/viewtopic.php?f=8&t=5&p=13704#p13704

Entry:
```
f=8&t=5&p=13704#p13704
```

MichaIng committed:
v6.13

+ DietPi-Software | Folding@Home: Add forum link

MichaIng committed:
v6.13

+ DietPi-Services | Add Folding@Home client to DietPi-Service control

MichaIng committed:
v6.13

+ DietPi-Software | Folding@Home: Fully pre-configure install, using default user name and passkey, which can be easily changed afterwards.
+ DietPi-Software | Folding@Home: Move data folder to dietpi_userdata to allow reduced SD writes.  Pre-create subfolder and symlink to let deb package directly install files to the right place. Carefully handle existing data and migrate in case.

MichaIng committed:
v6.13

+ DietPi-Software | Folding@Home: Add data folder to Install_Apply_Permissions
+ DietPi-Software | Folding@Home: Fix G_CONFIG_INJECT usage, escaping single quite does not work here (?)

MichaIng committed:
v6.13

+ DietPi-Software | Folding@Home: Wait 5 seconds, to assure FAHClient wrote down it's config file after stopping.

MichaIng commented:
🈴 Fails to start on Jessie. Verbose output: `Starting fahclient ... /usr/bin/FAHClient: /usr/lib/x86_64-linux-gnu/libstdc++.so.6: version 'GLIBCXX_3.4.21' not found (required by /usr/bin/FAHClient)`
- If we do not already apply a manual/backport update this particular library, I think we should drop Jessie support. Manually mess around with those library versions may otherwise break other software/system parts?

Ref: https://github.com/FoldingAtHome/fah-issues/issues/1147#issuecomment-134731211
- `libstdc++.so.6.0.21`+ required.
```
root@VM-Jessie:/tmp# l /usr/lib/x86_64-linux-gnu/ | grep libstdc++
lrwxrwxrwx 1 root root       19 Jan 31  2018 libstdc++.so.6 -> libstdc++.so.6.0.20
-rw-r--r-- 1 root root  1008120 Jan 31  2018 libstdc++.so.6.0.20
```
- 😝 one subversion too low.

Fourdee commented:
@MichaIng 

> If we do not already apply a manual/backport update this particular library, I think we should drop Jessie support. Manually mess around with those library versions may otherwise break other software/system parts?

Yep. The only Jessie image required is for Odroid's, due to Kodi support. Aside from that, we should be not be implementing any Jessie specific workarounds or support.

Fourdee commented:
Still work pending, but i've merged re: https://github.com/Fourdee/DietPi/issues/1985#issuecomment-410786747, and, testing purposes.

MichaIng commented:
@Fourdee 
Okay, so drop Jessie support and I would go for "Computational Science" now, which seems to fit best: https://en.wikipedia.org/wiki/Computational_science
**€: Done**

Testing needed, it also on slower SBCs to 5 seconds wait time, before adjusting `config.xml` is enough. When the service stops, it writes back the parsed configuration to this file, which takes a short time, after the service stop is already finished (does not wait for it).
Only after the `config.xml` is written back, we can edit it, otherwise our edits will be overwritten.

Fourdee commented:
@MichaIng 

If the config file does not exist until after 1st run is completed, you could try:
```
run_program

while true
do

    if [[ ! -f config.xml ]]; then

        sleep 2

    else
        
        stop program
        apply configuration changes
        break

    fi

done
```

MichaIng commented:
@Fourdee 
If I observed it right, the config.xml always exist. After deb install, a sample config.xml file is in place. On first start, the client recognizes that/if the file is untouched sample or not, and only in first case it creates (at first internally) a simplified/formatted new config based on the system (core count, GPU etc.).
This configuration is then written back/overwriting the sample config.xml when service fully stopped.

If service was started with sample config.xml in place, and then G_CONFIG_INJECT applies, on service stop all manual changes will be overwritten. If G_CONFIG_INJECT applies, before service fully started, then fahclient does not adjust it any more and will fail, as needed values are missing/wrong.

But the current method (waiting for full start/stop) should work? As far as I tested, service names are always the same and match the pattern.
